proposals/csharp-11.0/low-level-struct-improvements.md

```diff
@@ -869,6 +869,194 @@ ref struct S<T>
+The below examples make use of named lifetimes. The syntax `$a` refers to a lifetime named `a`. It is a lifetime that has no meaning by itself but can be given a relationship to other lifetimes via the `where lifetime` syntax. There are a few predefined lifetimes for convenience and brevity below:
+
+- `$heap`: this is the lifetime of any value that exists on the heap. It is available in all scopes and method signatures.
+- `$local`: this is the lifetime of any value that exists on the method stack. It's effectively a name place holder for *current method*. It is implicitly defined in methods and can appear in method signatures except for any output position.
+- `$ro`: name place holder for the *return only* safe to escape scope
+- `$cm`: name place holder for the *calling method* safe to escape scope
+
+There are a few predefined relationships between lifetimes:
+
+- `lifetime $heap >= $a` for all lifetimes `$a`
+- `lifetime $cm >= $ro` 
+- `lifetime $x >= $local` for all predefined lifetimes. User defined lifetimes have no relationship to local unless explictly defined.
+
+Lifetime variables when defined on types can be invariant or covariant. These are expressed using the same syntax as generic parameters:
+
+```csharp
+// $this is covariant
+// $a is invariant
+ref struct S<out $this, out $a> 
+```
+
+The lifetime of a ref is expressed by providing a lifetime argument to the ref. For example a `ref` that refers to the heap is expressed as `ref<$heap>`.
+
+When defining a constructor in the model the name `new` will be used for the method. It is necessary to have a parameter list for the returned value as well as the constructor arguments. This is necessary to express the relationship between constructor inputs and the constructed value. Rather than having `Span<$a><$ro>` the model will use `Span<$a> new<$ro>` instead. The type of `this` in the constructor, including lifetimes, will be the defined return value.
+
+The basic rules for the lifetime are defined as:
+
+- All lifetimes are expressed syntactically as generic arguments, coming before type arguments. This is true for predefined lifetimes except `$heap` and `$local`. 
+- All types `T` that are not a `ref struct` implicitly have lifetime of `T<$heap>`. This is implicit, there is no need to write `int<$heap>` in every sample.
+- For a ref field defined as `ref T<$l1, $l2, ... $ln>` all lifetimes `$l1` through `$ln` must be invariant. 
+- For a ref defined as `ref<$a> T<$b, ...>`, `$b` must be known to be at least as big as `$a`
+- The `ref` of a variable has a lifetime of the 
+    - For a ref local, parameter, field or return of type `ref<$a> T` the lifetime is `$a`
+    - `$heap` for all reference types and fields of reference types
+    - `$local` for everything else
+- An assignment or return is legal when the underlying type conversion is legal
+- A ref assignment or return where the LHS is `ref<$a> T1` and the RHS is `ref<$b> T2` is legal when both:
```

Comment:
> We should also define that a ref assignment with LHS ref<$a> readonly T1 and RHS ref<$b> readonly T2 is legal when T2 is a subtype of T1.

Yes that is an interesting case to put in there. Have to think about best language. 

> Also, it seems like it would be better if this were restated as a conversion rule, with no specific tie to returns, arguments, or assignments. One set of conversion rules applies in all places expressions are read.

I started writing it as a conversion rule but `ref` is not a type so my mind got stuck on whether this was or wasn't a conversion. Or rather how can we represent ref and non-ref assignment as different conversions if type system wise they're the same types since ref isn't part of the type?

> Let's also define the method lifetime argument inference rules, by saying that for each lifetime $la in the arguments which corresponds to a lifetime parameter $lp in the method's signature, a widest bound is set for the inferred argument for that parameter. 

I initially started to write inference rules but backed off. It felt like running before we learned to walk. Wanted to get everyone on the same page as what is the syntax and how do rules play out. Next step is okay now how do our language inference rules map down to these.

Reply:
> I started writing it as a conversion rule but ref is not a type so my mind got stuck on whether this was or wasn't a conversion. Or rather how can we represent ref and non-ref assignment as different conversions if type system wise they're the same types since ref isn't part of the type?

I think conversion rules make sense and would be nicely consistent with how we spec other parts of the language.

As for type system -- I'll propose we just spec refs as part of the type system for now. We don't need to actually make a "Ref<T>" type, just say "ref-ness is considered as part of the type compatibility in conversions".

My hope is we can end up with something precise enough that we all understand what it means and what's safe, while being lightweight enough that we don't get bogged down in specifying things that can never occur in C#.

Reply:
> ref-ness is considered as part of the type compatibility in conversions 

Think that works. Can avoid a giant spec change by just noting it is true for lifetime conversions. 

What I want to avoid is getting too tied into types such that it impacts our ability to implement. A 100% type based approach is not really feasible to implement given how Roslyn works today. Feeding extra generics through types: that's not practical. This is going to be implemented by a walker that is feeding back a state information per expression in the tree. So it's basically side car data we have to reconcile with the types involved. 

Don't really forsee any problems with the model we've described so far but we should keep it in mind when discussing things.